tests/reproducibility/fixtures.py

```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+
+def models_equals(model1, model2):
+    w1 = model1.get_weights()
+    w2 = model2.get_weights()
+    return all(np.array_equal(w, w_new) for w, w_new in zip(w1, w2))
+
+
+def assert_reproducible(func, num_iter=20):
+    """
+    Assert Keras models produced from calling ``func`` are reproducible.
+
+    Args:
+        func (callable): Function to check for reproducible model
+        num_iter (int, default 20): Number of iterations to run through to validate reproducibility.
+
+    """
+    model = func()
+    for i in range(num_iter):
+        model_new = func()
+        if not models_equals(model, model_new):
```

Comment:
Did you consider:

```suggestion
        assert models_equals(model, model_new)
```

I think pytest's introspect of assert statements should do a reasonable job of showing the two models: https://pytest.readthedocs.io/en/latest/assert.html#assert

Reply:
I'm guessing that only works when the assert is in the form `assert value1 == value2`? I tried a few alternatives but couldn't get it to print out anything about the model... which also led me to find the current form also just looks like:
```
>               assert False, (model, model_new)
E               AssertionError: (<tensorflow.python.keras.engine.training.Model object at 0x1432c6210>, <tensorflow.python.keras.engine.training.Model object at 0x13fb97990>)
```
so I might change it to be `assert False, (model.get_weights(), model_new.get_weights())`

Reply:
https://pytest.readthedocs.io/en/latest/example/reportingdemo.html#tbreportdemo shows various things that aren't just `assert a == b`, e.g. 

```
______________________ TestMoreErrors.test_startswith ______________________

self = <failure_demo.TestMoreErrors object at 0xdeadbeef>

    def test_startswith(self):
        s = "123"
        g = "456"
>       assert s.startswith(g)
E       AssertionError: assert False
E        +  where False = <built-in method startswith of str object at 0xdeadbeef>('456')
E        +    where <built-in method startswith of str object at 0xdeadbeef> = '123'.startswith

failure_demo.py:227: AssertionError
```

but maybe methods are special and functions aren't.

Instead of `assert False` (and the `assert True` later), what about:

```python
assert models_equals(model, model_new), (model.get_weights(), model_new.get_weights())
```

Doing `if f(): assert False` instead of `assert f()` seems, to me, similar to doing `True if f() else False` instead of just `f()`. 😄